src/main/java/com/thealgorithms/datastructures/disjointsetunion/DisjointSetUnion.java

```diff
@@ -0,0 +1,53 @@
+package com.thealgorithms.datastructures.disjointsetunion;
+
+/**
+ * Disjoint Set Union or DSU is useful for solving problems related to connected components,
+ * cycle detection in graphs, and maintaining relationships in disjoint sets of data.
+ * It is commonly employed in graph algorithms and problems.
+ *
+ * @see <a href="https://en.wikipedia.org/wiki/Disjoint-set_data_structure">Disjoint Set Union</a>
+ */
+public class DisjointSetUnion<T> {
+
+    /**
+     * Creates a new node of DSU with parent initialised as same node
+     */
+    public Node<T> makeSet(T x) {
+        return new Node<T>(x);
+    }
+
+    /**
+     * Finds and returns the representative (root) element of the set to which a given element belongs.
+     * This operation uses path compression to optimize future findSet operations.
+     */
+    public Node<T> findSet(Node<T> node) {
+        while (node != node.parent) {
+            node = node.parent;
+        }
+        return node;
+    }
+
+    /**
+     * Unions two sets by merging their representative elements. The merge is performed based on the rank of each set
+     * to ensure efficient merging and path compression to optimize future findSet operations.
+     */
+    public void unionSets(Node<T> x, Node<T> y) {
```

Comment:
```suggestion
    public void unionSets(final Node<T> x, final Node<T> y) {
```

Reply:
Updated testcases accordingly to check all branches
<img width="777" alt="Screenshot 2023-09-10 at 9 38 14 PM" src="https://github.com/TheAlgorithms/Java/assets/76104205/e28846ce-0505-440b-a849-0bc38ffd7a27">